Weighted pls -- adding support for sample_weight option.

Fenugreek opened the pull request:
Added support for sample_weight option, refactoring some code along the way. See commit message for more details.

I have tested that results do not change when no sample_weight option is specified, and when sample_weight = ones(len(X)). Also, in my weighted real-world use case, results look reasonable and are an improvement over using no weights.

Deepak

Commit:
Adding support to sample_weight option. In progress; incomplete.

Commit:
User visible changes
================

* Added support for sample_weight option
One weight per row of data can now be specified to the fit() functions via the sample_weight option.

* Algorithm option supported for PLSRegression
Previously, hardcoded as "nipals"; now can take on "svd" as well.
Also, in docstring usage notes swapped "inherited" for more appropriate "defaults to".

* self.coefs_ attribute and predict() method now supported only for PLSRegression
Not computed and not supported for PLSCanonical, PLSSVD and CCA as in these cases X and Y are interchangeable and the context is covarying latent factor finding rather than regression. See Wegelin paper referenced in the docstrings. This seems to have been the original intent of the original author (see FIXME comment in the code before my commit; this is my proposed resolution of the FIXME).

Other code changes
================

* moved input arrays check code out of fit() methods
So they are in one place, _check_inputs().

* _center_scale_xy() now supports sample_weight option
RFC: consider moving this (now more complex) function out of this model and into utils/stats.py?

* refactored deflation code in fit()
Now in separate function, _deflate(), called twice -- once for X and once for Y.

amueller commented:
Thanks for the contribution. Unfortunately I'm not really familiar enough with PLS to review the changes.
It seems after the refactoring the code got quite a bit longer. Is that mostly due to docs?

There are a bunch of test failures because some classes are now missing a ``predict`` method.

Commit:
Put back predict() functionality for classes other than PLSRegression --
otherwise, breaks tests. Should deprecate more gradually. Added comments to that effect.

Changed comment "Normalize" to "Center and scale".

Updated doctest to reflect new constructor option (algorithm=...).

Commit:
test new sample_weight option by supplying fit() method with sample_weight array
of identical weights, and checking that results do not change.

Fenugreek commented:
Thanks for the intial review, Andreas.

I've fixed the test failures by adding back the predict() method, and adding some comments to the code saying that this ought to be considered for deprecation. I also added a new test, which checks adding identical sample weights does not change results. Lastly, changed comment "Normalize" to "Center and scale" as you suggested, and also made the one-line docstring change you asked for.

As for reviewing the weight functionality itself, it boils down to reviewing the one line, line 86, which adds weights to the SVD, and one other location analagous to that for the NIPALS algorithm alternative, at lines 34-35, which is similar to how you do weighted linear least squares. If you're comfortable with those, I think we are good to go. If not or you don't think so, maybe I can write to the mailing list to ask for volunteers to review?

Thanks,
Deepak

Fenugreek commented:
Re: code getting longer, I think 90 lines longer, only
    _center_scale_xy is longer significantly. There are a bunch of
    places where I added a few lines each to handle sample weights.
    Also, I added a dozen or so empty lines.
    
    On 7/30/2015 1:07 PM, Andreas Mueller
      wrote:
    
    
      Thanks for the contribution. Unfortunately I'm not really
        familiar enough with PLS to review the changes.
        It seems after the refactoring the code got quite a bit longer.
        Is that mostly due to docs?
      There are a bunch of test failures because some classes are now
        missing a predict method.
      —
        Reply to this email directly or view
          it on GitHub.

Commit:
Merge remote-tracking branch 'upstream/master' into weighted-PLS

walterreade commented:
I'm not a contributor to sklearn, but I've read probably 90% of the papers on PLS. I can take a look at the math within the next 48 hours.

mblondel commented:
@walterreade That would be nice, thanks!

Fenugreek commented:
Hi Walter,
    
    do you have some time to take a look now? 
    There's also, a possibly more pressing, PLS issue, that probably
    someone who knows PLS can answer right away, here:
    https://github.com/scikit-learn/scikit-learn/issues/2821
    
    We really can use some PLS expertise :). 
      Deepak
    
    On 9/14/2015 2:11 PM, Walter Reade
      wrote:
    
    
      I'm not a contributor to sklearn, but I've read probably 90% of
        the papers on PLS. I can take a look at the math within the next
        48 hours.
      —
        Reply to this email directly or view
          it on GitHub.

walterreade commented:
So sorry! This dropped off my plate. I'll take a look this weekend. Thanks for the bump.

walterreade commented:
I'm currently working through the code changes. I'll probably have a list of thoughts, but the first question I had was regarding handling pathological weights (e.g., all zeros, negative weights). Is it normal to have these types of checks?

(Again, I'll apologize in advance for my any newbie questions.)

walterreade commented on sklearn/cross_decomposition/pls_.py:
```diff
@@ -72,40 +78,111 @@ def _nipals_twoblocks_inner_loop(X, Y, mode="A", max_iter=500, tol=1e-06,
     Returns
     -------
-        X, Y, x_mean, y_mean, x_std, y_std
+    X, Y, sample_weifht (if specified) as numpy arrays, of correct dimensions.
```
typo

Fenugreek commented:
Thanks Walter. Re: pathological weights, I had a look at other code in sklearn (e.g. logistic.py) and didn't see checks. Happy to add a check_weights() func to utils/validation.py and have this code call that if one of the core contributors suggests that's worthwhile.

Separately, thanks for the typo spot. Will fix along with other issues you raise.

walterreade commented:
Hi Deepak,

I can't comment on whether or not there should be a weights check (since I don't know the specific guidelines for sklearn contribution.) I was just pointing that out as a potential issue.

I've done a number of additional checks (e.g., weighting a single row to zero returns the same result of excluding that row from X and Y, etc.) I didn't not find anything amiss. I've tried to validate against an external algorithm, but there are so many pitfalls with that. So, unfortunately, I do not have anything concrete in that regard.

I don't have any other concerns, and personally think it will be a good addition. 

Lastly, sorry for the delayed responses. It's been a hectic few weeks.

Commit:
Fix typo in doc string.

Fenugreek commented:
Thanks Walter. I've now fixed the typo, and also created a new pull request, PR #5370 with a check_weights() utility function. Thanks again for your time.

Fenugreek commented:
(Sorry, user error in accidentally closing this PR. Reopened it.)

amueller commented:
@walterreade if you had some time to check our implementation against a reference, without sample weights or with, that would be really great :)

amueller commented on sklearn/cross_decomposition/tests/test_pls.py:
```diff
@@ -133,6 +133,17 @@ def check_ortho(M, err_msg):
+    assert_array_almost_equal(pls_2W.x_weights_, x_weights)
+    assert_array_almost_equal(pls_2W.x_loadings_, x_loadings)
+    assert_array_almost_equal(pls_2W.y_weights_, y_weights)
+    assert_array_almost_equal(pls_2W.y_loadings_, y_loadings)
```
can you add a test that a sample weight of zero is the same as removing a sample and that a sample weight of two is the same as duplicating a sample?

amueller commented:
I looked at it again and I think the changes look ok. A little bit more tests would be nice, though.

walterreade commented:
@amueller It will be out of town all next week, Sunday - Saturday, but when I'm back my current hectic schedule will have freed up. Then I will be able to focus on this again.

amueller commented:
Thank you, and don't worry. This has been suboptimal for a while and a week or two more doesn't really matter. Any help is much appreciated.

Commit:
New check_weights() function.

    Check if input array makes sense as sample weights:
    * all finite.
    * no negative value.
    * sum to greater than zero (this check can optionally be disabled).

Commit:
Call new check_weights() function in utils/validation.py to validate sample_weight.

Commit:
More tests for sample_weight option:
* Results should be same if sample weight is 0 or sample is removed.
* Results should be same if sample weight is double or sample is duped.

Fenugreek commented:
Thanks @amueller , @walterreade.  I have now added the weights sane values validation check (moved my PR #5370 to here), and also added the two tests @amueller suggested (zero weight same as removing sample, double weight same as duplicating sample) to tests/test_pls.py. 

make fails with some strange module import errors:
`ERROR: Failure: ValueError (Attempted relative import beyond toplevel package)`
so I am not sure if the two tests I added work fine, though I did try out the two situations (zero weight same as removing sample, double weight same as duplicating sample) with my own code and the results did match.

amueller commented:
where did you get the error? the tests should always use absolute imports (``from sklearn.X import X``)

Fenugreek commented:
Sorry, user error -- my local tree was somehow slightly corrupted. Now test runs OK, including the two new tests I'd added. 

(And I checked -- test_pls.py does use only absolute imports.)
